fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1256,268 @@ class SamOrder(enum.Enum):
+    ```
+
+    Format of a single secondary alignment in the `XB` tag (`,`-delimited):
+
+    ```text
+    chr,<orientation><1-based position>,cigar,NM,AS,MapQ
+    ```
+
+    Full example of an `XB` tag value (`;`-delimited):
+
+    ```text
+    XB:Z:chr9,-104599381,49M,4,0,30;chr3,+170653467,49M,4,0,20;chr12,+46991828,49M,5,0,10;
+    ```
+
+    Attributes:
+        SAM_TAGS: The SAM tags this class is capable of parsing.
+
+    Args:
+        reference_name: The reference sequence name.
+        reference_start: The 0-based start position of the alignment.
+        is_forward: If the alignment is in the forward orientation or not.
+        cigar: The Cigar sequence representing the alignment.
+        mapping_quality: The aligner-reported probability of an incorrect mapping, if available.
+        edit_distance: The number of mismatches between the query and the target, if available.
+        alignment_score: The aligner-reported alignment score, if available.
+        is_secondary: If this auxiliary alignment is a secondary alignment or not.
+        is_supplementary: If this auxiliary alignment is a supplementary alignment or not.
+
+    Raises:
+        ValueError: If `reference_start` is set to a value less than zero.
+        ValueError: If `mapping_quality` is set to a value less than zero.
+        ValueError: If `edit_distance` is set to a value less than zero.
+
+    See:
+        - [BWA User Manual](https://bio-bwa.sourceforge.net/bwa.shtml)
+        - [https://github.com/lh3/bwa/pull/292](https://github.com/lh3/bwa/pull/292)
+        - [https://github.com/lh3/bwa/pull/293](https://github.com/lh3/bwa/pull/293)
+        - [https://github.com/lh3/bwa/pull/355](https://github.com/lh3/bwa/pull/355)
+    """
+
+    SAM_TAGS: ClassVar[Collection[str]] = ("SA", "XA", "XB")
+
+    reference_name: str
+    reference_start: int
+    is_forward: bool
+    cigar: Cigar
+    mapping_quality: Optional[int] = None
+    edit_distance: Optional[int] = None
+    alignment_score: Optional[int] = None
+    is_secondary: bool = False
+    is_supplementary: bool = False
+
+    def __post_init__(self) -> None:
+        """Perform post-initialization validation on this dataclass."""
+        errors: list[str] = []
+        if self.reference_start < 0:
+            errors.append(f"Reference start cannot be less than 0! Found: {self.reference_start}")
+        if self.mapping_quality is not None and self.mapping_quality < 0:
+            errors.append(f"Mapping quality cannot be less than 0! Found: {self.mapping_quality}")
+        if self.edit_distance is not None and self.edit_distance < 0:
+            errors.append(f"Edit distance cannot be less than 0! Found: {self.edit_distance}")
+        # TODO: Some aligners might allow for a score <0 but I'm not sure bwa does... Keep this?
+        # if self.alignment_score is not None and self.alignment_score < 0:
+        #    errors.append(f"Alignment score cannot be less than 0! Found: {self.alignment_score}")
```

Comment:
@tfenne @nh13 there is a concern about keeping this validation.

Do you know what the right thing is to do?

Source: https://github.com/fulcrumgenomics/fgpyo/pull/208#discussion_r1910434919

Reply:
See the `bwa aln -T` parameter, which I have set negative to get more alignments.  I'd say remove it.